Lib/functools.py

```diff
@@ -945,8 +945,18 @@ def register(self, cls, method=None):
         return self.dispatcher.register(cls, func=method)
 
     def __get__(self, obj, cls=None):
-        if self._all_weakrefable_instances and cls is not None and obj in self._method_cache:
-            return self._method_cache[obj]
+        caching = self._all_weakrefable_instances and obj is not None
+        cache = self._method_cache
+
+        if caching:
+            try:
+                _method = cache[obj]
+            except TypeError:
+                self._all_weakrefable_instances = caching = False
+            except KeyError:
+                pass
+            else:
+                return _method
```

Comment:
```suggestion
        cache = self._method_cache

        if self._all_weakrefable_instances and obj is not None:
            try:
                _method = cache[obj]
            except TypeError:
                self._all_weakrefable_instances = caching = False
            except KeyError:
                caching = True
            else:
                return _method
```
This avoids assigning a local variable `caching` in the fast path. One can argue about the readability

Reply:
```pytb
======================================================================
ERROR: test_staticmethod_type_ann_register (test.test_functools.TestSingleDispatch.test_staticmethod_type_ann_register)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\alexw\coding\fast-cpython\Lib\test\test_functools.py", line 2634, in test_staticmethod_type_ann_register
    self.assertTrue(A.t(0))
                    ^^^
  File "C:\Users\alexw\coding\fast-cpython\Lib\functools.py", line 966, in __get__
    if caching:
       ^^^^^^^
UnboundLocalError: cannot access local variable 'caching' where it is not associated with a value

----------------------------------------------------------------------
Ran 259 tests in 0.187s

FAILED (errors=10)
test test_functools failed
test_functools failed (10 errors)

== Tests result: FAILURE ==

1 test failed:
    test_functools

Total duration: 313 ms
Tests result: FAILURE
```

Reply:
ah, it's because we don't assign `caching` if we don't hit the `if self._all_weakrefable_instances and obj is not None` branch